pkg/rancher/migrations.go

```diff
@@ -0,0 +1,75 @@
+package rancher
+
+import (
+	"github.com/mcuadros/go-version"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	rancherversion "github.com/rancher/rancher/pkg/version"
+	controllerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	cattleNamespace          = "cattle-system"
+	forceUpgradeLogoutConfig = "forceupgradelogout"
+	rancherVersionKey        = "rancherVersion"
+)
+
+// forceUpgradeLogout will delete all dashboard tokens forcing a logout.  This is useful when there is a major frontend
+// upgrade and we want all users to be sent to a central point.  This function will check for the `forceUpgradeLogoutConfig`
+// configuration map and only run if the last migrated version is lower than the given `migrationVersion`.
+func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, tokenController v3.TokenController, migrationVersion string) error {
+	cm, err := configMapController.Cache().Get(cattleNamespace, forceUpgradeLogoutConfig)
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+
+	// if this is the first ever migration initialize the configmap
+	if cm == nil {
+		cm = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      forceUpgradeLogoutConfig,
+				Namespace: cattleNamespace,
+			},
+			Data: map[string]string{},
```

Comment:
Maps should be initialized with `make()` unless they have data as part of that initialization

Reply:
@dramich , isn't this purely preference? https://stackoverflow.com/a/31064737/16564280
There are 100+ places in rancher/rancher code that initialize empty map with `map[string]string{}` syntax.

Reply:
It's a rancher coding standard and best practice, I don't review every PR that goes into rancher